File: apps/server/src/modules/reports/reports.service.ts
import { OrderStatus } from '@prisma/client';
import { prisma } from '../../lib/prisma';

export type Period = 'today' | 'week' | 'month' | 'year';

function getDateRange(period: Period) {
  const now = new Date();
  const from = new Date();
  if (period === 'today') { from.setHours(0, 0, 0, 0); }
  else if (period === 'week') { from.setDate(now.getDate() - 7); }
  else if (period === 'month') { from.setMonth(now.getMonth() - 1); }
  else { from.setFullYear(now.getFullYear() - 1); }
  return { from, to: now };
}

export async function getSummary(merchantId: string, period: Period) {
  const { from, to } = getDateRange(period);

  const [total, delivered, pending, cancelled, processing, inDelivery] =
    await prisma.$transaction([
      prisma.order.count({ where: { merchantId, createdAt: { gte: from, lte: to } } }),
      prisma.order.count({ where: { merchantId, status: OrderStatus.DELIVERED, createdAt: { gte: from, lte: to } } }),
      prisma.order.count({ where: { merchantId, status: OrderStatus.PENDING, createdAt: { gte: from, lte: to } } }),
      prisma.order.count({ where: { merchantId, status: OrderStatus.CANCELLED, createdAt: { gte: from, lte: to } } }),
      prisma.order.count({ where: { merchantId, status: OrderStatus.PROCESSING, createdAt: { gte: from, lte: to } } }),
      prisma.order.count({ where: { merchantId, status: OrderStatus.IN_DELIVERY, createdAt: { gte: from, lte: to } } }),
    ]);

  const successRate = total > 0 ? Math.round((delivered / total) * 100) : 0;

  return { total, delivered, pending, cancelled, processing, inDelivery, successRate, period, from, to };
}

export async function getOrdersChart(merchantId: string, period: Period) {
  const { from, to } = getDateRange(period);
  const orders = await prisma.order.findMany({
    where: { merchantId, createdAt: { gte: from, lte: to } },
    select: { createdAt: true, status: true },
    orderBy: { createdAt: 'asc' },
  });

  // Group by date label
  const map = new Map<string, { delivered: number; cancelled: number; total: number }>();
  for (const o of orders) {
    const key = o.createdAt.toISOString().slice(0, 10);
    const existing = map.get(key) ?? { delivered: 0, cancelled: 0, total: 0 };
    existing.total++;
    if (o.status === OrderStatus.DELIVERED) existing.delivered++;
    if (o.status === OrderStatus.CANCELLED) existing.cancelled++;
    map.set(key, existing);
  }

  return Array.from(map.entries()).map(([date, counts]) => ({ date, ...counts }));
}

export async function getAreaStats(merchantId: string, period: Period) {
  const { from, to } = getDateRange(period);
  const orders = await prisma.order.groupBy({
    by: ['governorate'],
    where: { merchantId, createdAt: { gte: from, lte: to } },
    _count: { id: true },
    orderBy: { _count: { id: 'desc' } },
  });

  return orders.map((r) => ({ governorate: r.governorate, count: r._count.id }));
}

export async function getDriverStats(merchantId: string, period: Period) {
  const { from, to } = getDateRange(period);
  const orders = await prisma.order.groupBy({
    by: ['driverName'],
    where: {
      merchantId,
      driverName: { not: null },
      createdAt: { gte: from, lte: to },
    },
    _count: { id: true },
    orderBy: { _count: { id: 'desc' } },
  });

  return orders
    .filter((r) => r.driverName)
    .map((r) => ({ driver: r.driverName!, count: r._count.id }));
}

export async function getTimeStats(merchantId: string, period: Period) {
  const { from, to } = getDateRange(period);
  const orders = await prisma.order.findMany({
    where: { merchantId, createdAt: { gte: from, lte: to } },
    select: { createdAt: true },
  });

  const hours = Array.from({ length: 24 }, (_, i) => ({ hour: i, count: 0 }));
  for (const o of orders) {
    hours[o.createdAt.getHours()].count++;
  }
  return hours;
}

File: apps/merchant-app/lib/services/reports_service.dart
import 'api_client.dart';

enum ReportPeriod { today, week, month, year }

extension ReportPeriodX on ReportPeriod {
  String get apiValue => switch (this) {
        ReportPeriod.today => 'today',
        ReportPeriod.week => 'week',
        ReportPeriod.month => 'month',
        ReportPeriod.year => 'year',
      };
}

class ReportSummary {
  final int total;
  final int delivered;
  final int pending;
  final int cancelled;
  final int processing;
  final int inDelivery;
  final int successRate;

  const ReportSummary({
    required this.total,
    required this.delivered,
    required this.pending,
    required this.cancelled,
    required this.processing,
    required this.inDelivery,
    required this.successRate,
  });

  factory ReportSummary.fromJson(Map<String, dynamic> j) => ReportSummary(
        total: j['total'] as int? ?? 0,
        delivered: j['delivered'] as int? ?? 0,
        pending: j['pending'] as int? ?? 0,
        cancelled: j['cancelled'] as int? ?? 0,
        processing: j['processing'] as int? ?? 0,
        inDelivery: j['inDelivery'] as int? ?? 0,
        successRate: j['successRate'] as int? ?? 0,
      );

  static ReportSummary empty() => const ReportSummary(
        total: 0,
        delivered: 0,
        pending: 0,
        cancelled: 0,
        processing: 0,
        inDelivery: 0,
        successRate: 0,
      );
}

class ChartPoint {
  final String date; // "2024-12-15"
  final int total;
  final int delivered;
  final int cancelled;

  const ChartPoint({
    required this.date,
    required this.total,
    required this.delivered,
    required this.cancelled,
  });

  factory ChartPoint.fromJson(Map<String, dynamic> j) => ChartPoint(
        date: j['date'] as String,
        total: j['total'] as int? ?? 0,
        delivered: j['delivered'] as int? ?? 0,
        cancelled: j['cancelled'] as int? ?? 0,
      );
}

class AreaStat {
  final String governorate;
  final int count;
  const AreaStat({required this.governorate, required this.count});

  factory AreaStat.fromJson(Map<String, dynamic> j) => AreaStat(
        governorate: j['governorate'] as String? ?? '',
        count: j['count'] as int? ?? 0,
      );
}

class ReportsService {
  ReportsService._();
  static final ReportsService instance = ReportsService._();

  final _api = ApiClient.instance;

  Future<ReportSummary> getSummary(ReportPeriod period) async {
    final res = await _api.get('/reports/summary?period=${period.apiValue}');
    return ReportSummary.fromJson(res);
  }

  Future<List<ChartPoint>> getOrdersChart(ReportPeriod period) async {
    final res = await _api.getList('/reports/orders-chart?period=${period.apiValue}');
    return res.map((e) => ChartPoint.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<AreaStat>> getAreaStats(ReportPeriod period) async {
    final res = await _api.getList('/reports/areas?period=${period.apiValue}');
    return res.map((e) => AreaStat.fromJson(e as Map<String, dynamic>)).toList();
  }
}
